In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm

import warnings
warnings.filterwarnings("ignore")

In [4]:
import pandas as pd
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

1. Önceki derste olduğu gibi, hedef değişkeninizin sıcaklık ve görünür sıcaklık arasındaki fark olduğu bir doğrusal regresyon modeli oluşturun. Açıklayıcı değişkenler olarak nem ve rüzgar hızı kullanın. Şimdi, modelinizi OLS kullanarak tahmin edin. R-kare ve ayarlanmış R-kare değerleri nelerdir? Tatminkar olduklarını düşünüyor musunuz? Açıklayın.

In [19]:
X = weather_df[["Humidity","Wind Speed (km/h)"]]
Y = weather_df["Temperature (C)"] - weather_df["Apparent Temperature (C)"]
lrm = LinearRegression()
lrm.fit(X, Y)

x = sm.add_constant(X)
results = sm.OLS(Y, x).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Tue, 06 Jul 2021   Prob (F-statistic):               0.00
Time:                        13:53:13   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- The r squared and adjusted r squared values are 0.288 for two of them. These values are quite low. It will be good for us if it is at least over 0.70. Too much may mean overfitting.

2. Daha sonra, yukarıdaki modele nem ve rüzgar hızı etkileşimini dahil edin ve OLS'yi kullanarak modeli tahmin edin. Şimdi, bu modelin R-kare değeri nedir? Bu model bir öncekine göre gelişti mi?

In [17]:
weather_df["Hum_wind"] = weather_df["Humidity"] * weather_df["Wind Speed (km/h)"]
new_X = X = weather_df[["Humidity","Wind Speed (km/h)","Hum_wind"]]

lrm = LinearRegression()
lrm.fit(new_X, Y)

x = sm.add_constant(new_X)
results = sm.OLS(Y, x).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Tue, 06 Jul 2021   Prob (F-statistic):               0.00
Time:                        13:52:55   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity             -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)    -0.0905      0.002    -36.797      0.000      -0.095      -0.086
Hum_wind              0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- The R squared value of the model is 0.341 It has made some progress.

3. İlk modele ek açıklayıcı değişken olarak görünürlük ekleyin ve tahmin edin. R-kare arttı mı? Ayarlanmış R-kare değeri ne oldu? Tabloda ortaya çıkan farklılıkları, ayarlanmış R-kare içindeki iyileşme açısından etkileşim terimi ve görünürlük ile karşılaştırın. Hangisi daha kullanışlı?

In [18]:
X = weather_df[["Humidity","Wind Speed (km/h)","Visibility (km)"]]

lrm = LinearRegression()
lrm.fit(X, Y)

print("Temperature (C)")
x = sm.add_constant(X)
results = sm.OLS(Y, x).fit()
results.summary()

Temperature (C)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Tue, 06 Jul 2021   Prob (F-statistic):               0.00
Time:                        13:52:59   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -1.5756      0.028    -56.605      0.000      -1.630      -1.521
Humidity              2.6066      0.025    102.784      0.000       2.557       2.656
Wind Speed (km/h)     0.1199      0.001    179.014      0.000       0.119       0.121
Visibility (km)      -0.0540      0.001    -46.614      0.000      -0.056      -0.052
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.279
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                           0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- R square and R square selections of our model have been progressed. It's less than the model we built with 
interaction term.

4. AIC ve BIC puanlarına göre yukarıdaki üç modelden en iyisini seçin. Mentor ile gerekçenizi tartışarak seçiminizi doğrulayın.

The best of these three models is the 2nd model. The lower the AIC and BIC values, the better for us.

In [23]:
df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

1. Ev fiyatları modelinizi tekrar çalıştırın ve F-testi, R-kare, ayarlanmış R-kare, AIC ve BIC kullanarak modelinizin uygunluğunu değerlendirin.

In [25]:
dummies = pd.get_dummies(df[["SaleType","Street"]],drop_first=True)
X = pd.concat([df["FullBath"],dummies],axis=1,ignore_index=False)
Y = df["SalePrice"]

lrm = LinearRegression()
lrm.fit(X, Y)

x = sm.add_constant(X)
results = sm.OLS(Y, x).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.373
Model:                            OLS   Adj. R-squared:                  0.369
Method:                 Least Squares   F-statistic:                     86.33
Date:                Tue, 06 Jul 2021   Prob (F-statistic):          1.99e-139
Time:                        14:04:22   Log-Likelihood:                -18203.
No. Observations:                1460   AIC:                         3.643e+04
Df Residuals:                    1449   BIC:                         3.649e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const           3.077e+04    2.8e+04      1.101      0.271   -2.41e+04    8.56e+04
FullBath        7.217e+04   3095.223     23.318      0.000    6.61e+04    7.82e+04
SaleType_CWD    4.103e+04    3.3e+04      1.243      0.214   -2.37e+04    1.06e+05
SaleType_Con    1.181e+05   4.56e+04      2.587      0.010    2.85e+04    2.08e+05
SaleType_ConLD -7347.2825   2.33e+04     -0.315      0.753   -5.31e+04    3.84e+04
SaleType_ConLI  4.165e+04   2.98e+04      1.396      0.163   -1.69e+04       1e+05
SaleType_ConLw  -608.9474   2.98e+04     -0.020      0.984   -5.91e+04    5.79e+04
SaleType_New    8.733e+04   1.13e+04      7.697      0.000    6.51e+04     1.1e+05
SaleType_Oth   -1.965e+04   3.77e+04     -0.521      0.602   -9.36e+04    5.43e+04
SaleType_WD     1.973e+04   9794.824      2.015      0.044     518.159    3.89e+04
Street_Pave      1.25e+04   2.61e+04      0.479      0.632   -3.86e+04    6.36e+04
==============================================================================
Omnibus:                      567.515   Durbin-Watson:                   1.998
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3827.742
Skew:                           1.657   Prob(JB):                         0.00
Kurtosis:                      10.207   Cond. No.                         65.6
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- The model's R squared, adjusted R squared, and F test values ​​are very low, which is not what we want. For AIC and BIC, the less the better, but we see that the values ​​are very high.

2. Modelinizin tatmin edici olduğunu düşünüyor musunuz? Açıklayın

- Our model is certainly not satisfactory. As I said in the previous question, the values ​​are not at the level we want.

3. Modelinizin uygunluğunu iyileştirmek için, bazı değişkenleri ekleyerek veya kaldırarak farklı model özelliklerini deneyin.

In [35]:
new_df = df[["FullBath","MSSubClass","LotFrontage","MoSold","SalePrice"]].dropna()
X = new_df[["FullBath","MSSubClass","LotFrontage","MoSold"]]
Y = new_df["SalePrice"].dropna()

lrm = LinearRegression()
lrm.fit(X, Y)

x = sm.add_constant(X)
results = sm.OLS(Y, x).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.387
Model:                            OLS   Adj. R-squared:                  0.385
Method:                 Least Squares   F-statistic:                     188.9
Date:                Tue, 06 Jul 2021   Prob (F-statistic):          1.57e-125
Time:                        14:11:01   Log-Likelihood:                -15018.
No. Observations:                1201   AIC:                         3.005e+04
Df Residuals:                    1196   BIC:                         3.007e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const        1.062e+04   9334.800      1.138      0.256   -7694.897    2.89e+04
FullBath     8.099e+04   3599.701     22.500      0.000    7.39e+04    8.81e+04
MSSubClass   -169.5717     48.825     -3.473      0.001    -265.363     -73.780
LotFrontage   725.7829     87.612      8.284      0.000     553.893     897.673
MoSold        329.9197    700.395      0.471      0.638   -1044.220    1704.059
==============================================================================
Omnibus:                      462.061   Durbin-Watson:                   2.035
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2605.221
Skew:                           1.688   Prob(JB):                         0.00
Kurtosis:                       9.377   Cond. No.                         473.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [40]:
new_df = df[["FullBath","MSSubClass","LotFrontage","MoSold","SalePrice"]].dropna()
X = new_df[["FullBath","MSSubClass","LotFrontage"]]
Y = new_df["SalePrice"].dropna()

lrm = LinearRegression()
lrm.fit(X, Y)

x = sm.add_constant(X)
results = sm.OLS(Y, x).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.387
Model:                            OLS   Adj. R-squared:                  0.386
Method:                 Least Squares   F-statistic:                     252.0
Date:                Tue, 06 Jul 2021   Prob (F-statistic):          1.01e-126
Time:                        14:13:02   Log-Likelihood:                -15019.
No. Observations:                1201   AIC:                         3.005e+04
Df Residuals:                    1197   BIC:                         3.007e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const        1.262e+04   8310.821      1.518      0.129   -3686.079    2.89e+04
FullBath     8.112e+04   3588.174     22.608      0.000    7.41e+04    8.82e+04
MSSubClass   -170.4935     48.770     -3.496      0.000    -266.177     -74.810
LotFrontage   724.9820     87.567      8.279      0.000     553.181     896.783
==============================================================================
Omnibus:                      460.515   Durbin-Watson:                   2.035
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2578.713
Skew:                           1.684   Prob(JB):                         0.00
Kurtosis:                       9.339   Cond. No.                         423.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- Among these three models, the last one is the best because although the other values ​​are the same, the F statistic value is larger.